test/solver_tests/ilp_solver_test.py

```diff
@@ -313,16 +313,34 @@ def test_ilp_solver_perfect_phylogeny(self):
         multi_parents = [n for n in tree if tree.in_degree(n) > 1]
         self.assertEqual(len(multi_parents), 0)
 
+        # make sure the resulting tree has no unifurcations
+        one_child = [n for n in tree if tree.out_degree(n) == 1]
+        self.assertEqual(len(one_child), 0)
+
         # expected parsimony
         expected_parsimony = 6
         root = roots[0]
 
         observed_parsimony = 0
         for e in nx.dfs_edges(tree, source=root):
-            if tree.out_degree(e[1]) > 0:
-                observed_parsimony += cas.solver.dissimilarity.hamming_distance(
-                    e[0], e[1]
-                )
+            if tree.out_degree(e[0]) > 0:
+                if e[1] in expected_leaves:
+                    observed_parsimony += (
+                        cas.solver.dissimilarity.hamming_distance(
+                            e[0],
+                            str(
+                                tuple(
+                                    self.duplicates_tree.character_matrix.loc[
+                                        e[1]
+                                    ]
+                                )
+                            ),
+                        )
+                    )
+                else:
```

Comment:
I don't think we need all of this logic. If this is because we have hanging leaves on the ILPSolver that aren't part of the sample input, then we actually have a small bug on our hands and should amend the solver post-processing code.

Reply:
hi Matt,
This was added because after collapsing unifurcations on the final ILP reconstructed tree we need a new calculation of the parsimony, as we need to add the parsimony from the leaves which are now not attached to an internal node representing their character vector.

Reply:
I've fixed this in a new push to the PR.